man/zoxide.1
.TH "ZOXIDE" "1" "2021-04-12" "" "zoxide"
.SH NAME
\fBzoxide\fR - a smarter cd command
.SH SYNOPSIS
.B zoxide SUBCOMMAND [OPTIONS]
.SH DESCRIPTION
zoxide is a smarter cd command for your terminal. It keeps track of the
directories you use most frequently, and uses a ranking algorithm to navigate
to the best match.
.SH USAGE
.nf
$ z foo        # cd into highest ranked directory matching foo
$ z foo bar    # cd into highest ranked directory matching foo and bar
.sp
$ z ~/foo      # z also works like a regular cd command
$ z foo/       # cd into relative path
$ z ..         # cd one level up
$ z -          # cd into previous directory
.sp
$ zi foo       # cd with interactive selection (using fzf)
.fi
.SH SUBCOMMANDS
.TP
\fBzoxide-add\fR(1)
Add a new directory to the database, or increment its rank.
.TP
\fBzoxide-import\fR(1)
Import entries from another application.
.TP
\fBzoxide-init\fR(1)
Generate shell configuration.
.TP
\fBzoxide-query\fR(1)
Search for a directory in the database.
.TP
\fBzoxide-remove\fR(1)
Remove a directory from the database.
.SH OPTIONS
.TP
.B -h, --help
Print help information.
.TP
.B -V, --version
Print version information.
.SH ENVIRONMENT VARIABLES
Environment variables can be used for configuration. They must be set before
\fBzoxide-init\fR(1) is called.
.TP
.B _ZO_DATA_DIR
Specifies the directory in which the database is stored. The default value
varies across OSes:
.TS
tab(|);
l l.
    \fBOS|Path\fR
    \fBLinux/BSD\fR|T{
\fB$XDG_DATA_HOME\fR or \fB$HOME/.local/share\fR, eg.
\fB/home/alice/.local/share\fR
T}
    \fBmacOS\fR|T{
\fB$HOME/Library/Application Support\fR, eg.
\fB/Users/Alice/Library/Application Support\fR
T}
    \fBWindows\fR|T{
\fB{FOLDERID_RoamingAppData}\fR, eg. \fBC:\\Users\\Alice\\AppData\\Roaming\fR
T}
.TE
.TP
.B _ZO_ECHO
When set to 1, \fBz\fR will print the matched directory before navigating to it.
.TP
.B _ZO_EXCLUDE_DIRS
Prevents the specified directories from being added to the database. This is
provided as a list of globs, separated by OS-specific characters:
.TS
tab(|);
l l.
    \fBOS|Separator\fR
    \fBLinux/macOS/BSD\fR|T{
\fB:\fR, eg. \fB$HOME:$HOME/private/*\fR
T}
    \fBWindows\fR|\fB;\fR, eg. \fB$HOME;$HOME/private/*\fR
.TE
.sp
By default, this is set to \fB$HOME\fR. After setting this up, you might need
to use \fBzoxide-remove\fR(1) to remove any existing entries from the database.
.TP
.B _ZO_FZF_OPTS
Custom options to pass to \fBfzf\fR(1) during interactive selection. See the
manpage for the full list of options.
.TP
.B _ZO_MAXAGE
Configures the aging algorithm, which limits the maximum number of entries in
the database. By default, this is set to 10000.
.TP
.B _ZO_RESOLVE_SYMLINKS
When set to 1, \fBz\fR will resolve symlinks before adding directories to
the database.
.SH ALGORITHM
.TP
.B AGING
zoxide uses a parameter called \fB_ZO_MAXAGE\fR to limit the number of entries
in the database based on usage patterns. If the total \fBFRECENCY\fR of the
directories in the database exceeds this value, we divide each directory's
score by a factor \fBk\fR - such that the new total becomes ~90% of
\fB_ZO_MAXAGE\fR. Thereafter, if the new score of any directory falls below
1, it is removed from the database.
.sp
Theoretically, the maximum number of directories in the database is
\fB4 * _ZO_MAXAGE\fR, although it is lower in practice.
.TP
.B FRECENCY
Each directory in zoxide is given a score, starting with 1 the first time
it is accessed. Every subsequent access increases the score by 1. When a
query is made, we calculate frecency based on the last time the directory was
accessed:
.TS
tab(|);
l l.
    \fBLast access time\fR|\fBFrecency\fR
    Within the last hour|score * 4
    Within the last day|score * 2
    Within the last week|score / 2
    Otherwise|score / 4
.TE
.SH REPORTING BUGS
For any issues, feature requests, or questions, please visit:
.sp
\fBhttps://github.com/ajeetdsouza/zoxide/issues\fR.
.SH AUTHOR
Ajeet D'Souza <\fB98ajeet@gmail.com\fR>


CHANGELOG.md
<!-- markdownlint-disable-file MD024 -->

# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## Unreleased

### Changed

- manpages: moved to `man/man1/*.1`.

## [0.8.0] - 2021-12-25

### Added

- Zsh: completions for `z` command.

### Changed

- Fzf: better default options.
- Fish: interactive completions are only triggered when the last argument is empty.
- PowerShell: installation instructions.

### Fixed

- PowerShell: use global scope for aliases.
- Zsh: fix errors with `set -eu`.
- Fzf: handle early selection.
- PowerShell: correctly handle escape characters in paths.
- Parse error on Cygwin/MSYS due to CRLF line endings.
- Fzf: handle spaces correctly in preview window.
- Bash: avoid initializing completions on older versions.
- Fzf: avoid launching binary from current directory on Windows.

## [0.7.9] - 2021-11-02

### Changed

- Bash/Fish: improved completions for `z` command.

### Fixed

- Fish: error erasing completions on older versions.
- PowerShell: enable `--cmd cd` to replace the `cd` command.

## [0.7.8] - 2021-10-21

### Added

- Auto-generated completions for [Fig](https://fig.io/).

### Fixed

- Compile error with `clap v3.0.0-beta.5`.

## [0.7.7] - 2021-10-15

### Fixed

- PowerShell: Hook not initializing correctly.

## [0.7.6] - 2021-10-13

### Changed

- Nushell: upgrade minimum supported version to v0.37.0.

### Fixed

- Xonsh: error messages in `zi`.
- Xonsh: configuration environment variables not being handled correctly.

## [0.7.5] - 2021-09-09

### Added

- Bash/Elvish: completions for `z` command.

### Changed

- Nushell: upgrade minimum supported version to v0.36.0.
- Nushell: easier installation instructions.

### Fixed

- Elvish: unable to `z` into directories by path.
- Elvish: don't show traceback when `z` or `zi` fails.
- Elvish: nested shells do not initialize correctly.

## [0.7.4] - 2021-08-15

### Fixed

- Compile error with `clap v3.0.0-beta.4`.

## [0.7.3] - 2021-08-05

### Added

- `zoxide add` and `zoxide remove` now accept multiple arguments.

### Fixed

- Nushell: errors on 0.33.0.
- PowerShell: errors when initializing in `StrictMode`.
- Bash/POSIX: remove conflicting alias definitions when initializing.
- Bash: remove extra semicolon when setting `$PROMPT_COMMAND`.
- Xonsh: use shell environment instead of `os.environ`.

## [0.7.2] - 2021-06-10

### Fixed

- `zoxide -V` not printing version.

## [0.7.1] - 2021-06-09

### Added

- Auto-generated shell completions.
- `zoxide query --all` for listing deleted directories.
- Lazy deletion for removed directories that have not been accessed in > 90 days.
- Nushell: support for 0.32.0+.

### Fixed

- Nushell: avoid calling `__zoxide_hook` on non-filesystem subshells.
- Fish: `alias cd=z` now works, but it must be done after calling `zoxide init`.
- PowerShell: avoid calling `__zoxide_hook` on non-filesystem providers.
- Fish: avoid calling `__zoxide_hook` in private mode.

## [0.7.0] - 2021-05-02

### Added

- Manpages for all subcommands.
- Default prompt for Nushell.

### Changed

- `zoxide remove -i` now accepts multiple selections.
- `zoxide add` no longer accepts zero parameters.
- `$_ZO_EXCLUDE_DIRS` now defaults to `"$HOME"`.
- Binary releases now use `.zip` on Windows, `.tar.gz` otherwise.

### Fixed

- `cd -` on fish shells.
- `__zoxide_hook` no longer changes value of `$?` within `$PROMPT_COMMAND` on bash.

### Removed

- GitHub install script.
- Release binaries built with `glibc`, use `musl` instead.

## [0.6.0] - 2021-04-09

### Added

- Support for [Nushell](https://www.nushell.sh/).
- Support for [Elvish](https://elv.sh/).

### Changed

- `z` now excludes the current directory from search results.

### Fixed

- Removed backtraces on Rust nightly.
- Generated shell code avoids using aliased builtins.
- Handle broken pipe errors gracefully when writing to streams.
- NUL file appearing in working directory on Windows.
- Accidental redefinition of hooks when initialized twice on some shells.
- zoxide unable to find itself on Xonsh shells.

### Removed

- Aliases: `za`, `zq`, `zqi`, `zr`, `zri`. These are trivial aliases that can
  easily be defined manually, and aren't very useful to most users.

## [0.5.0] - 2020-10-30

### Added

- `$_ZO_EXCLUDE_DIRS` now supports globs.
- `zoxide init` now defines `__zoxide_z*` functions that can be aliased as needed.
- Support for the [xonsh](https://xon.sh/) shell.
- `zoxide import` can now import from Autojump.

### Changed

- `zoxide init --no-aliases` no longer generates `z` or `zi`.

### Fixed

- Clobber conflicting alias definitions in bash/fish/zsh/POSIX shells.

### Removed

- Deprecated PWD hooks for POSIX shells.
- Lazy deletion for inaccessible directories.

## [0.4.3] - 2020-07-04

### Fixed

- Bug in Fish init script

## [0.4.2] - 2020-07-03

### Added

- `$_ZO_FZF_OPTS` to specify custom options for `fzf`
- `zoxide query --list` to list all matches
- `zoxide query --score` to show score along with result

### Changed

- Increased default value of `$_ZO_MAXAGE` to `10000`.
- Symlinks are treated as separate directories by default, this can be changed
  by setting `_ZO_RESOLVE_SYMLINKS=1`.

### Removed

- Help menus for `z` and `zri`.
- `zoxide remove -i` is replaced with `zri`.

## [0.4.1] - 2020-05-25

### Added

- Support for powershell.

### Removed

- Backward compatibility with `v0.2.x` databases.
- Support for paths with invalid UTF-8.

## [0.4.0] - 2020-05-03

### Added

- Interactive mode for removing entries (`zoxide remove -i`).
- Aliases for interactive `query` and `remove` (`zqi` and `zri` respectively).
- PWD hooks for POSIX shells.

### Changed

- `zoxide remove` now throws an error if there was no match in the database.
- Interactive mode in `zoxide` no longer errors out if `fzf` exits gracefully.

contrib/completions/zoxide.ts
const completion: Fig.Spec = {
  name: "zoxide",
  description: "A smarter cd command for your terminal",
  subcommands: [
    {
      name: "add",
      description: "Add a new directory or increment its rank",
      options: [
        {
          name: ["-h", "--help"],
          description: "Print help information",
        },
        {
          name: ["-V", "--version"],
          description: "Print version information",
        },
      ],
      args: {
        name: "paths",
        isVariadic: true,
        template: "folders",
      },
    },
    {
      name: "import",
      description: "Import entries from another application",
      options: [
        {
          name: "--from",
          description: "Application to import from",
          args: {
            name: "from",
            suggestions: [
              {
                name: "autojump",
              },
              {
                name: "z",
              },
            ]
          },
        },
        {
          name: "--merge",
          description: "Merge into existing database",
        },
        {
          name: ["-h", "--help"],
          description: "Print help information",
        },
        {
          name: ["-V", "--version"],
          description: "Print version information",
        },
      ],
      args: {
        name: "path",
        template: "filepaths",
      },
    },
    {
      name: "init",
      description: "Generate shell configuration",
      options: [
        {
          name: "--cmd",
          description: "Renames the 'z' command and corresponding aliases",
          args: {
            name: "cmd",
            isOptional: true,
          },
        },
        {
          name: "--hook",
          description: "Chooses event upon which an entry is added to the database",
          args: {
            name: "hook",
            isOptional: true,
            suggestions: [
              {
                name: "none",
              },
              {
                name: "prompt",
              },
              {
                name: "pwd",
              },
            ]
          },
        },
        {
          name: "--no-aliases",
          description: "Prevents zoxide from defining any commands",
        },
        {
          name: ["-h", "--help"],
          description: "Print help information",
        },
        {
          name: ["-V", "--version"],
          description: "Print version information",
        },
      ],
      args: {
        name: "shell",
        suggestions: [
          {
            name: "bash",
          },
          {
            name: "elvish",
          },
          {
            name: "fish",
          },
          {
            name: "nushell",
          },
          {
            name: "posix",
          },
          {
            name: "powershell",
          },
          {
            name: "xonsh",
          },
          {
            name: "zsh",
          },
        ]
      },
    },
    {
      name: "query",
      description: "Search for a directory in the database",
      options: [
        {
          name: "--exclude",
          description: "Exclude a path from results",
          args: {
            name: "exclude",
            isOptional: true,
            template: "folders",
          },
        },
        {
          name: "--all",
          description: "Show deleted directories",
        },
        {
          name: ["-i", "--interactive"],
          description: "Use interactive selection",
        },
        {
          name: ["-l", "--list"],
          description: "List all matching directories",
        },
        {
          name: ["-s", "--score"],
          description: "Print score with results",
        },
        {
          name: ["-h", "--help"],
          description: "Print help information",
        },
        {
          name: ["-V", "--version"],
          description: "Print version information",
        },
      ],
      args: {
        name: "keywords",
        isOptional: true,
      },
    },
    {
      name: "remove",
      description: "Remove a directory from the database",
      options: [
        {
          name: ["-i", "--interactive"],
          description: "Use interactive selection",
        },
        {
          name: ["-h", "--help"],
          description: "Print help information",
        },
        {
          name: ["-V", "--version"],
          description: "Print version information",
        },
      ],
      args: {
        name: "paths",
        isOptional: true,
        template: "folders",
      },
    },
  ],
  options: [
    {
      name: ["-h", "--help"],
      description: "Print help information",
    },
    {
      name: ["-V", "--version"],
      description: "Print version information",
    },
  ],
};

export default completion;
